### ISSUE TITLE
Unable to make transfers

### ISSUE THREAD

**Issue by CosminNechifor** [raiden-network/raiden#3190]
## Issue checklist

To help us more easily debug your issue please check the boxes for which requirements for safe usage or Raiden were followed:

- [X] The Ethereum node was always up to date with the chain tip and never stopped running.
- [X] No Ethereum transaction was sent for the account that Raiden manages by another app.
- [X] The Ethereum account used by Raiden always had sufficient ETH.
- [X] The Raiden DB that exists by default at ~/.raiden was never altered or deleted.
- [ ] The Raiden node was never restarted. If yes please provide more info.
Was restarted a few time by me, but it was always given enough time to recover.
- [X] The ethereum node was not switched while Raiden was running.
- [X] The Raiden REST API is protected from public access.

-->

## Problem Definition

I tried to perform a transfer to @kelsos using my long running node but it didn't work. After that I opened a channel with @czepluch and tried to perform the transfer through him, the result was the same. The channel I had with @kelsos was closed by him and we ended up with the following topology:

```
C <-> J <-> K
```
I have to mention that I was able to perform payments to Jacob using our channel, but I couldn't send Kelsos any.

If I was using the WebUI to send payments it would provide no feedback/response and if I tried to perfom the payment using curl it would start hanging:
``curl -i -X POST http://localhost:5001/api/v1/payments/0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2/0x4d6FA60E64C7E520899A9aeDD40cB1E6992027F0 -H 'Content-Type: application/json' --data-raw '{"amount": 1}`` 

### Logs and database:

Cosmin logs: [node_b2c94bba.zip](https://github.com/raiden-network/raiden/files/2695706/node_b2c94bba.zip)

### System
```
Linux cosmin 4.14.83-1-MANJARO #1 SMP PREEMPT Fri Nov 23 18:20:11 UTC 2018 x86_64 GNU/Linux
```

**Comment by kelsos** [comment]
I have the original logs [here](https://drive.google.com/open?id=1CzgoVAfGF2z51GGQAO7kQ9Eb-54x0nHZ) from the direct channel if there is need I can also provide the logs from later.

I was also trying to send to Cosmin through the WebUI and it seems that all of my attempts would timeout on the WebUI

**Comment by rakanalh** [comment]
@czepluch Can you please provide logs & databases?

**Comment by czepluch** [comment]
[raiden-debug-jacob.zip](https://github.com/raiden-network/raiden/files/2698481/raiden-debug-jacob.zip)

**Comment by rakanalh** [comment]
Thanks guys,

@CosminNechifor it would have been nice if you added more details from our logs to the description. The word hanging means that we have a nonce mismatch somewhere so this piece of information being logged somewhere and included in this issue could have helped me debug this alot, compared to going through the logs of 3 nodes with so many transfers going back and forth.

**Comment by CosminNechifor** [comment]
Will pay more attention to that from now on Rakan

**Comment by rakanalh** [comment]
Reminder to self:
Jacob's node received a lock expired from Kelsos' with nonce 653:
```
> ReceiveLockExpired
        sender='0x4d6fa60e64c7e520899a9aedd40cb1e6992027f0'
        balance_proof=<BalanceProofSignedState nonce:653 transferred_amount:265 locked_amount:1 locksroot:52bfddbb token_network:a5c9ecf5 channel_identifier:83 message_hash:b056bf51 signature:4254288b sender:4d6fa60e chain_id:1>
        secrethash='0x0cd8968dd7bc3541d9c78b95b2e116c6b1babbc278482531dea8294579f61dda'
        message_identifier=5649877514784376478
```
But at block `6915868`, the task was removed because the lock expired. So the above message wasn't processed the nonce mismatch happened.
I should check where the node for Kelsos sent this message and see

**Comment by rakanalh** [comment]
Investigation Result:

I will refer to:
0xf2fBCC167cc3f24a620798E27b4BB084dD275253 as A
0x4d6fa60e64c7e520899a9aedd40cb1e6992027f0 as B

Initially, A sends B a transfer to mediate
```
> ActionInitMediator
        sender='0xf2fbcc167cc3f24a620798e27b4bb084dd275253'
        balance_proof=<BalanceProofSignedState nonce:358 transferred_amount:111 locked_amount:9 locksroot:f4033ea9 token_network:a5c9ecf5 channel_identifier:83 message_hash:023794fc signature:5d4d3432 sender:f2fbcc16 chain_id:1>
        routes=[<RouteState hop:f0daca3b channel_identifier:82>]
        from_route=<RouteState hop:f2fbcc16 channel_identifier:83>
        from_transfer=<LockedTransferSignedState msgid:7156235094316858804 id:1545235637186 token:0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2 lock:<HashTimeLockState amount:1 expiration:6915858 secrethash:0cd8968d> target:0x4bfa865bddb97fddc871ab8bddf390e293ebd015>
< SendProcessed
        recipient='0xf2fbcc167cc3f24a620798e27b4bb084dd275253'
        queue_identifier=<QueueIdentifier recipient:f2fbcc16 channel_identifier:0>
        message_identifier=7156235094316858804
< SendLockedTransfer
        recipient='0xf0daca3b7cb562059d89bd5d8291a482b20f9d52'
        queue_identifier=<QueueIdentifier recipient:f0daca3b channel_identifier:82>
        message_identifier=3706708253869010486
        transfer=<LockedTransferUnsignedState id:1545235637186 token:0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2 balance_proof:<BalanceProofUnsignedState nonce:170 transferred_amount:73 locked_amount:2 locksroot:89c5aa2b token_network:a5c9ecf5 channel_identifier:82 chain_id:1> lock:<HashTimeLockState amount:1 expiration:6915858 secrethash:0cd8968d> target:0x4bfa865bddb97fddc871ab8bddf390e293ebd015>
```

B receives a refund as the next hop couldn't forward the transfer so B refunds A
```
> ReceiveTransferRefund
        sender='0xf0daca3b7cb562059d89bd5d8291a482b20f9d52'
        balance_proof=<BalanceProofSignedState nonce:558 transferred_amount:272 locked_amount:2 locksroot:89c5aa2b token_network:a5c9ecf5 channel_identifier:82 message_hash:b165bc22 signature:ebdbc849 sender:f0daca3b chain_id:1>
        transfer=<LockedTransferSignedState msgid:6851114587851295491 id:1545235637186 token:0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2 lock:<HashTimeLockState amount:1 expiration:6915858 secrethash:0cd8968d> target:0x4bfa865bddb97fddc871ab8bddf390e293ebd015>
        routes=[<RouteState hop:f2fbcc16 channel_identifier:83>]
< SendProcessed
        recipient='0xf0daca3b7cb562059d89bd5d8291a482b20f9d52'
        queue_identifier=<QueueIdentifier recipient:f0daca3b channel_identifier:0>
        message_identifier=6851114587851295491
< SendRefundTransfer
        recipient='0xf2fbcc167cc3f24a620798e27b4bb084dd275253'
        queue_identifier=<QueueIdentifier recipient:f2fbcc16 channel_identifier:83>
        message_identifier=6662618582127026459
        transfer=<LockedTransferUnsignedState id:1545235637186 token:0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2 balance_proof:<BalanceProofUnsignedState nonce:649 transferred_amount:264 locked_amount:1 locksroot:15ee2d11 token_network:a5c9ecf5 channel_identifier:83 chain_id:1> lock:<HashTimeLockState amount:1 expiration:6915858 secrethash:0cd8968d> target:0x4bfa865bddb97fddc871ab8bddf390e293ebd015>
```

So A in this case tries a new route. According to the following code:
https://github.com/raiden-network/raiden/blob/f5d48172931ce60c820253e81708acc2a2f49941/raiden/transfer/mediated_transfer/initiator_manager.py#L238-L244

A creates a new sub-iteration which would eventually change the transfer's description, therefore changing the transfer:
https://github.com/raiden-network/raiden/blob/f5d48172931ce60c820253e81708acc2a2f49941/raiden/transfer/mediated_transfer/initiator.py#L223-L228

At the following block:
```
> Block
        block_number=6915868
        gas_limit=8000029
        block_hash='0x07099b82f404b7b90df438152c6cec959dcb93439a13eec28de9398531be9f3d'
< SendLockExpired
        recipient='0xb2c94bbaf307c6dfbf70cc2537385553385a345b'
        queue_identifier=<QueueIdentifier recipient:b2c94bba channel_identifier:88>
        message_identifier=3093397297185992673
        balance_proof=<BalanceProofUnsignedState nonce:171 transferred_amount:80 locked_amount:1 locksroot:52bfddbb token_network:a5c9ecf5 channel_identifier:88 chain_id:1>
        secrethash='0xbe6b3a35c1c447cfcc3c8db043eb8688dde448e268ab6423440d8cfc69604bc8'
< EventPaymentSentFailed
        payment_network_identifier='0xb9633dd9a9a71f22c933bf121d7a22008f66b908'
        token_network_identifier='0xa5c9ecf54790334b73e5dfa1ff5668eb425dc474'
        identifier=1545235637186
        target='0x4bfa865bddb97fddc871ab8bddf390e293ebd015'
        reason="transfer's lock has expired"
```
One of the locks in the payment task expires, and because we don't find any matching locks in `partner_state`, we clear the task. However, notice that the secret hash `0xbe6b3a35c1c447cfcc3c8db043eb8688dde448e268ab6423440d8cfc69604bc8` is different than the original transfer's secret hash: `0x0cd8968dd7bc3541d9c78b95b2e116c6b1babbc278482531dea8294579f61dda` and that is caused by the fact that we initiated a re-route transfer operation which lead the transfer task with secrethash `0x0cd8968dd7bc3541d9c78b95b2e116c6b1babbc278482531dea8294579f61dda` to actually contain a transfer with another secrethash (latest re-routed transfer).

At the same block, B sends A a lock expired message for the refund transfer:
```
> Block
        block_number=6915868
        gas_limit=8000029
        block_hash='0x07099b82f404b7b90df438152c6cec959dcb93439a13eec28de9398531be9f3d'
< EventUnlockClaimFailed
        identifier=1545235637186
        secrethash='0x0cd8968dd7bc3541d9c78b95b2e116c6b1babbc278482531dea8294579f61dda'
        reason='lock expired'
< EventUnlockClaimFailed
        identifier=1545235637186
        secrethash='0x0cd8968dd7bc3541d9c78b95b2e116c6b1babbc278482531dea8294579f61dda'
        reason='lock expired'
< SendLockExpired
        recipient='0xf0daca3b7cb562059d89bd5d8291a482b20f9d52'
        queue_identifier=<QueueIdentifier recipient:f0daca3b channel_identifier:82>
        message_identifier=536853268578252114
        balance_proof=<BalanceProofUnsignedState nonce:174 transferred_amount:74 locked_amount:0 locksroot:00000000 token_network:a5c9ecf5 channel_identifier:82 chain_id:1>
        secrethash='0x0cd8968dd7bc3541d9c78b95b2e116c6b1babbc278482531dea8294579f61dda'
```
Which fails because in the payment task whose secrethash is `0x0cd8968dd7bc3541d9c78b95b2e116c6b1babbc278482531dea8294579f61dda` was removed.

This issue is caused by the fact that the the initiator task did not have all transfers use the same secrethash which is assumed in several parts of the code base in initiator. 

This issue is potentially resolved by: https://github.com/raiden-network/raiden/issues/2329

**Comment by LefterisJP** [comment]
> This issue is potentially resolved by: #2329

I think that with a proper implementation of the above issue then this issue will also be fixed yes.

The required tasks to handle this issue would be:

> 1. Add a list of initiator_transfers to the InitiatorPaymentState
> 2. Subdispatch relevant state changes to all such initiator transfers of an InitiatorPaymentState

**Comment by rakanalh** [comment]
Thanks for confirming, will start progress on this

**Comment by rakanalh** [comment]
Should be resolved with #3227
